p4-16/spec/P4-16-spec.mdk

```diff
@@ -3819,6 +3820,33 @@ table tbl {
 }
 ~ End P4Example
 
+# Function declarations { #sec-functions }
+
+Functions can be declared only at the top-level.  Functions are very
+similar to functions in languages such as C or Java.  However, P4 does
+not permit recursive functions.
+
+~ Begin P4Grammar
+functionDeclaration
+    : functionPrototype blockStatement
+    ;
+~ End P4Grammar
+
+Here is an example of a function that returns the maximum of two 32-bit values:
+
+~ Begin P4Example
+bit<32> max(in bit<32> left, in bit<32> right) {
+   if (left > right)
+      return left;
+   return right;
+}
+~ End P4Example
+
+A function returns a value using the `return` statement.  A function
+that returns `void` can simply use the `return` statement with no
+arguments.  A function with a non-void return type must return a value
+of the suitable type on all possible execution paths.
```

Comment:
Should we also discuss that it is only visible to the data plane, and therefore all arguments must be directional?

Reply:
Actually non-directional arguments would work: they have to be bound to compile-time constants. We may want to prohibit them, because you can use them to pass externs or controls as arguments.

Reply:
In the interest of converging, I'd recommend that we stick with the conservative choice for now...